File: my-folder/0130-surrounded-regions/solution.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        rows, cols = len(board), len(board[0])

        def dfs(r,c):
            if (r<0 or r>=rows or c<0 or c>=cols or board[r][c]!='O'):
                return
            board[r][c] = 'T'
            dfs(r+1,c)
            dfs(r-1,c)
            dfs(r,c+1)
            dfs(r,c-1)
        
        for r in range(rows):
            for c in range(cols):
                if ( board[r][c]== 'O' and  (r in [0,rows -1] or c in [0, cols-1]) ):
                    dfs(r,c)
        # replace all O with X
        for i in range(rows):
            for j in range(cols):
                if board[i][j]=='O':
                    board[i][j] = 'X'
        
        for i in range(rows):
            for j in range(cols):
                if board[i][j] == 'T':
                    board[i][j] = 'O'
        

File: my-folder/0130-surrounded-regions/solution.cpp
class Solution {
public:
    void dfs(int r, int c, vector<vector<char>>& board)
        {
            int rows = board.size();
            int cols = board[0].size();
            if  ( (r< 0 or r>=rows) or (c<0 or c>=cols) or (board[r][c]!='O') ) return;
            board[r][c] = 'T';
            dfs(r+1,c, board);
            dfs(r-1,c,board);
            dfs(r,c+1,board);
            dfs(r,c-1,board);
        }

    void solve(vector<vector<char>>& board) {
        
        int rows = board.size();
        int cols = board[0].size();
        
        
        for (int i = 0; i<rows;i++)
        {
            for (int j = 0; j<cols; j++)
            {
                if ( (i==0 or i ==rows -1 or j==0 or j==cols -1 ) and (board[i][j]=='O') )
                {
                    dfs(i,j,board);
                }
            }
        }

        for (int i= 0; i< rows; i++)
        {
            for (int j = 0; j<cols;j++)
            {
                if (board[i][j]== 'O')
                {
                    board[i][j] = 'X';
                }
            }
        }

        for (int i= 0; i< rows; i++)
        {
            for (int j = 0; j<cols;j++)
            {
                if (board[i][j]== 'T')
                {
                    board[i][j] = 'O';
                }
            }
        }
        
    }
};
